### ISSUE TITLE
Perf: ContentFileUtils uses expensive file globbing when not needed

### ISSUE THREAD

**Issue by zivkan** [NuGet/Home#12695]
### NuGet Product Used

Other/NA

### Product Version

all

### Impact

None

### Repro Steps & Context

This codepath is used when restoring projects using PackageReference, and uses packages that contain `contentFiles/`. Therefore, restore performance could be improved.

When a package contains `contentFiles/`, and the project uses PackageReference, `ContentFileUtils.cs` is used to match the files in the zip, to metadata in the `.nuspec`. This code currently [always uses a file globbing API to do the match](https://github.com/NuGet/NuGet.Client/blob/4b6a87bfe90d247f961eba41d63dbf418a56fdcc/src/NuGet.Core/NuGet.Commands/RestoreCommand/ContentFiles/ContentFileUtils.cs#L104-L148). However, using [NuGet Insights](https://github.com/NuGet/Insights), we can see that most packages that have `contentFiles/` do not use `*` or `?` characters in the `.nuspec` `<contentFiles>` definitions.

Debugging restore with a suitable package, the way the NuGet uses the file globbing library, it will see a file `contentFiles/any/any/something.txt`, and a "pattern" `any/any/something.txt`, and the file globbing library will ask for the "root directory" (which in this case is `contentFiles/`, then ask if `contentFiles/` contains `any/`, then ask if `contentFiles/any/` contains `any/`, and finally if `contentFiles/any/any/` contains `something.txt`.

So, as we can see, there's a lot of unnecessary work that needs to be done because NuGet has a list of strings that represent the full path of all content files in the zip, and it's usually trying to match the substring following the `contentFiles/` prefix.

Therefore, I believe this could be implemented to be much higher performance, although there are some `/` vs `\` matching, and case sensitivity, issues that need to be accounted for.

Basically, do "simple" string matching when the nuspec's entry's include doesn't contain either `*`, or `?` (and maybe if there's also no `exclude`, but an exclude doesn't make sense on an exact filename). Only use the file globbing library when file globbing is actually needed.

**Comment by davkean** [comment]
@zivkan Had a thought, what if NuGet install of a package resolved resolved globs filters and turned them into direct references to the files it copied into the package directory, so it never needs to do this during restore? Alternatively could do it at pack time but that wouldn't benefit old packages.

**Comment by zivkan** [comment]
> Had a thought, what if NuGet install of a package resolved resolved globs filters and turned them into direct references to the files it copied into the package directory, so it never needs to do this during restore?

While NuGet doesn't currently have anything to validate that the files on disk match the files in the nupkg (for example to detect when people using hard links accidentally overwrite files in the global packages folder), I have in the past suggested to customers to write scripts to do so, helping them find these hard link and symbolic link issues in their builds. Of course, if NuGet added such validation, it could re-use the "expand contentFiles in the nuspec" logic in the comparison, but it will make 3rd party tools more difficult (and will break all existing scripts/tools).

All up, it'd have to be a trade off between ease of validating file corruption vs how many packages use file globbing patterns and how many files those globs actually match.

> Alternatively could do it at pack time but that wouldn't benefit old packages.

Even if it doesn't benefit old packages, if it nudges all new packages to use the low-cost path, so it's still a benefit.

In either case, I'm wondering if a glob that matches many files might actually have a negative impact if expanded. If there is 1 glob matching 1000 files, that's 1000 iterations of the current loop, but will be 1,000,000 iterations if the glob is expanded.
